lib/std/os.zig

```diff
@@ -438,6 +438,39 @@ pub fn pread(fd: fd_t, buf: []u8, offset: u64) PReadError!usize {
     return index;
 }
 
+pub const TruncateError = error{
+    /// The file descriptor is not open for writing.
+    NotFile,
+} || UnexpectedError;
+
+pub fn truncate(fd: fd_t, length: u64) TruncateError!void {
+    if (std.Target.current.os.tag == .windows) {
+        try windows.SetFilePointerEx_BEGIN(fd, length);
+
+        if (windows.kernel32.SetEndOfFile(fd) == 0)
+            return TruncateError.Unexpected;
+
+        return;
+    }
+
+    while (true) {
+        const rc = if (builtin.link_libc) blk: {
+            if (std.Target.current.os.tag == .linux)
+                break :blk system.ftruncate64(fd, @bitCast(off_t, length))
+            else
+                break :blk system.ftruncate(fd, @bitCast(off_t, length));
+        } else
+            system.ftruncate(fd, length);
+
+        switch (errno(rc)) {
+            0 => return,
+            EINTR => continue,
+            EBADF, EINVAL => return error.NotFile,
```

Comment:
I think it would be better to have these be `unreachable`. Although it's technically possible to devise a situation where you would rely on the kernel returning EBADF or EINVAL to detect a file descriptor opened in the wrong mode, it's almost certainly a bug or race condition if either of these occur. I would argue that there is no use case for having these as error set values, since correct and optimized software can be written to avoid these errors. Also, there is precedence for this in the rest of os.zig, so if you want to change things you'll have to open a separate proposal.

Reply:
Hitting `unreachable` is less pleasant that getting an error stating what's wrong, but I'll change it in a follow up commit.